Add Button component

domhorvathtw opened the pull request:
## Initial Button component

![image](https://user-images.githubusercontent.com/47105236/53487796-6c628200-3a84-11e9-9125-a24744cce755.png)
![image](https://user-images.githubusercontent.com/47105236/53488351-f4955700-3a85-11e9-80cc-72eae5e66eab.png)
![image](https://user-images.githubusercontent.com/47105236/53488393-0545cd00-3a86-11e9-89f8-e8a36920d579.png)
 
---

### Following design specs
https://transferwise.atlassian.net/wiki/spaces/DS/pages/859604203/Buttons

### Variants
- Primary
- Send
- Default
- Error

### Sates
- Default
- Loading 
- Disabled

### Different sizes 
- Extra small 
- Small 
- Medium (default) 
- Large

### Block option
Added isBlock flag

### Colours
Colour variables should get updated in bootstrap variables
https://github.com/transferwise/bootstrap/blob/master/less/variables.less

**Get new brand colour codes here:** 
https://app.goabstract.com/share/38662c17-b93d-454a-8a45-ea53b92e58f6

### Button layout variants
Circle, Dropdown and Split Button variants will be implemented in separate PR.

dozortsev commented on src/button/Button.docs.js:
```diff
@@ -0,0 +1,138 @@
+      variant: 'pay',
+      size: 'md',
+      block: false,
+      onClick: () => null,
```
I would add some action here e.g `alert(..)` (we use it in a few other component's docs)

domhorvathtw replied:
Added alert

dozortsev commented on src/button/Button.docs.js:
```diff
@@ -0,0 +1,138 @@
+                    selected={stateSelection}
+                    options={STATES.map(state => ({
+                      value: state,
+                      label: !state ? 'default' : state,
```
please avoid using `!`, plus this statement could be simplified `state && 'default'`

domhorvathtw replied:
that's actually completely unecessary, good catch, thanks!

dozortsev commented on src/button/Button.js:
```diff
@@ -0,0 +1,51 @@
+
+export const BUTTON_VARIANTS = ['primary', 'pay', 'secondary', 'danger'];
+export const BUTTON_SIZES = ['xs', 'sm', 'md', 'lg'];
+export const BUTTON_STATES = ['default', 'loading', 'disabled'];
```
why we need expose this constants?

domhorvathtw replied:
good catch, they were used in documentation as well

domhorvathtw commented on src/button/Button.docs.js:
```diff
@@ -0,0 +1,138 @@
+
+  render() {
+    const { label, stateSelection, size, variant, block, onClick } = this.state;
+    const buttonLabel = label === '' ? 'Send Money' : label;
```
fixed, thanks!

domhorvathtw commented:
The `<button>` element supports pressing enter when it has the focus. Select it with the tab and press enter and you'll see it works. The submit example is not necessary in my opinion, that's a specific use case, that you can achieve by using the button with the provided props (variants and states).

StevePole commented:
Could you make it so the text doesn't move when in loading state?

![image](https://user-images.githubusercontent.com/1318111/57301233-6b774d80-70d0-11e9-8e14-e3a4a2460aaf.png)

StevePole commented:
Could you make it so the text doesn't move when loading?

![button-loading](https://user-images.githubusercontent.com/1318111/57301267-834ed180-70d0-11e9-8dd3-accf5eec982c.gif)

domhorvathtw commented:
> Could you make it so the text doesn't move when loading?
> 
> ![button-loading](https://user-images.githubusercontent.com/1318111/57301267-834ed180-70d0-11e9-8dd3-accf5eec982c.gif)

Good catch, would have been annoying to have the text jumping back and forth between state changes. Since I added custom CSS, I also fixed the loader position and sizing for `xs` buttons.

oliverviljamaa commented on src/button/index.js:
```diff
@@ -0,0 +1,3 @@
+import Button from './Button';
```
How about `export { default } from './Button';`?

domhorvathtw replied:
fixed 🙇

oliverviljamaa commented on src/button/Button.js:
```diff
@@ -0,0 +1,52 @@
+
+Button.propTypes = {
+  variant: Types.oneOf(BUTTON_VARIANTS),
+  size: Types.oneOf(BUTTON_SIZES),
```
How about exposing the `Variant`, `Size`, and `State` as enums for consumers?

So I could do:
```js
<Button variant={ButtonVariant.PAY} size={ButtonSize.SMALL} buttonState={ButtonState.DISABLED}>
```

oliverviljamaa replied:
We could use the same enum-like objects to avoid repetition and more fragile string checks in this file as well 😊

domhorvathtw replied:
Agree 👍

domhorvathtw replied:
Agree 👍  exposed the enum-like objects `ButtonVariants`, `ButtonStates`, `ButtonSizes`

oliverviljamaa commented on src/button/Button.js:
```diff
@@ -0,0 +1,52 @@
+      disabled={isDisabled || buttonState === 'loading'}
+    >
+      {label}
+      {isLoading && <span className={`btn-loader ${block ? '' : 'm-l-2'}`} />}
```
What would you think about using `classNames` here?
```suggestion
      {isLoading && <span className={classNames('btn-loader', { 'm-l-2': block })} />}
```

domhorvathtw replied:
fixed

oliverviljamaa commented on src/button/Button.docs.js:
```diff
@@ -0,0 +1,142 @@
+
+  render() {
+    const { label, stateSelection, size, variant, block, onClick } = this.state;
+    const buttonLabel = label || 'Send Money';
```
Do we want to have a default in the demo? I found that a bit unintuitive, but it may just be me 😀

domhorvathtw replied:
do you mean not having a default label like "Send Money"?
if we don't have it, how would the button look like initially?

oliverviljamaa replied:
It's already in the state, so by default, it would say `Send money`. However, if we remove it from here (`render`), then the button will be empty when I erase the word from the docs input, which is expected, imo.

domhorvathtw replied:
True, removed it ✅

oliverviljamaa commented on src/button/Button.docs.js:
```diff
@@ -0,0 +1,142 @@
+export default class ButtonDocs extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Could we define `state` as a class field declaration and avoid the constructor here?

domhorvathtw replied:
I don't have a strong preference about state initialisation? I quickly checked, we use both constructor and class property to initialise state in components and component's docs. I feel like it's another discussion about creating / following a company wide react code style, what do you think?

oliverviljamaa replied:
You're right, it's something that should be enforced by ESLint 🙂 I normally just prefer the shorter version, so class field declarations, but there's really no big difference, so let's leave it as is here.

oliverviljamaa commented on src/button/Button.docs.js:
```diff
@@ -0,0 +1,142 @@
+            </div>
+            <div className="col-md-6">
+              <div className="row">
+                <div className="col-md-12">
```
If Bootstrap grid is 12 by default and this class makes it 12 from `md` onwards, it seems that it's just always 12 and we can remove the `div.row` and the `div.col-md-12`. But it's possible I'm just looking at it wrong.

domhorvathtw replied:
nope, you're right, my bad. Fixed it!

oliverviljamaa commented on src/button/Button.js:
```diff
@@ -0,0 +1,68 @@
+import classNames from 'classnames';
+import './Button.less';
+
+export const ButtonVariants = {
```
Thanks for adding these!
As they're enum-like structures and enums are usually named in singular, what do you think of having them in singular here as well?

https://stackoverflow.com/questions/15755955/naming-of-enums-in-java-singular-or-plural

domhorvathtw replied:
Sure thing, makes sense

oliverviljamaa commented on src/button/Button.js:
```diff
@@ -0,0 +1,68 @@
+  Disabled: 'disabled',
+};
+
+const Button = ({ label, buttonState, size, variant, block, onClick }) => {
```
I may have missed some conversations about it, but is there a reason why we've gone with `variant` here rather than `type`, for example?

I feel _variant_ is something that _differs_ from some norm, but for us, every button type is equal, even the default button has a type.

domhorvathtw replied:
Not really, I don't want to stick to `variant`. Let's use `type` and `variant` could be used when a button is different in shape or functionality, e.g. Circular, Dropdown, Split buttons. What do you think?

oliverviljamaa replied:
Agreed, I couldn't even think of the added benefit of being able to use `variant` in future scenarios!

oliverviljamaa commented on src/button/Button.spec.js:
```diff
@@ -0,0 +1,58 @@
+    wrapper.simulate('click');
+    expect(props.onClick).toBeCalled();
+  });
+});
```
Great tests. 👌

domhorvathtw replied:
👍
